### ISSUE TITLE
Docker PS exponentially slower with number of links in the system

### ISSUE THREAD

**Issue by tbatchelli** [docker/docker#9967]
I have been hitting this issue where building a site composed of many microservices ~20 becomes unbearably slow (~8 minutes). The same behavior can be seen on VirtualBox and in rackspace servers.

The docker daemon becomes exponentially slower, eventually taking 100% of one core for longer and longer (minutes at a time!) as new containers are created. This behavior does NOT correlate with number of containers created NOR with the number of images in the host. Instead it correlates with the number of links between all the existing containers (running or stopped)

To expose this issue, I have built a script that generates N containers running 'sleep 10000' on busybox (virtually cpu-less once started). Each of these containers is linked to each and everyone of the previously created containers.

```bash
#!/bin/bash
function links(){
    c=1
    while [ $c -lt $1 ]
    do
	echo -n "--link=test-$c:test-$c "
	c=$(( $c + 1 ))
    done
}

function run_test() {
    for i in  $(eval echo "{1..$1}")
     do
	echo "Starting Container #$i: docker run -d --name test-$i $(links $i) busybox sleep 10000"
	docker run -d --name test-$i $(links $i) busybox sleep 10000
	echo "docker ps -a | wc -l $(docker ps -a | wc -l)"
     done
}

# the following wraps run test to print the current time in seconds for each line in the output
(echo "Starting Container Creation";
 run_test $1 2>&1;
 echo "Done") | while read -r line; do echo "$(date +%s) $line"; done 
```

The following is the summary of the running the script above with 22 containers and exhibits the behavior I am describing: the first iteration is done in 1s or less, and the last iteration takes __412s or close to 7 min__. Notice that each iteration of the loop creates a container and then runs 'docker ps -a':

```
[~] ➔ cat output.txt | grep "docker ps"
docker ps -a
1420533414 docker ps -a | wc -l        2
1420533415 docker ps -a | wc -l        3
1420533416 docker ps -a | wc -l        4
1420533416 docker ps -a | wc -l        5
1420533417 docker ps -a | wc -l        6
1420533418 docker ps -a | wc -l        7
1420533419 docker ps -a | wc -l        8
1420533419 docker ps -a | wc -l        9
1420533420 docker ps -a | wc -l       10
1420533421 docker ps -a | wc -l       11
1420533422 docker ps -a | wc -l       12
1420533423 docker ps -a | wc -l       13
1420533425 docker ps -a | wc -l       14
1420533427 docker ps -a | wc -l       15
1420533431 docker ps -a | wc -l       16
1420533438 docker ps -a | wc -l       17
1420533451 docker ps -a | wc -l       18
1420533476 docker ps -a | wc -l       19
1420533527 docker ps -a | wc -l       20
1420533627 docker ps -a | wc -l       21
1420533831 docker ps -a | wc -l       22
1420534243 docker ps -a | wc -l       23
```

What's most relevant is that the creation of a container takes constant time in every iteration, but it is the call to 'docker ps -a' that takes exponentially longer to finish. This section of the the output of the script above shows shows the last two iterations and you can see by the timestamps that 'docker ps' is the culprit:

```
1420533627 docker ps -a | wc -l       21
1420533627 Starting Container #21: docker run -d --name test-21 --link=test-1:test-1 --link=test-2:test-2 --link=test-3:test-3 --link=test-4:test-4 --link=test-5:test-5 --link=test-6:test-6 --link=test-7:test-7 --link=test-8:test-8 --link=test-9:test-9 --link=test-10:test-10 --link=test-11:test-11 --link=test-12:test-12 --link=test-13:test-13 --link=test-14:test-14 --link=test-15:test-15 --link=test-16:test-16 --link=test-17:test-17 --link=test-18:test-18 --link=test-19:test-19 --link=test-20:test-20  busybox sleep 10000
1420533627 2f1336b330a0096524ccc888303d6b33b618f172d0c875209b892c25e41456d3
1420533831 docker ps -a | wc -l       22
1420533831 Starting Container #22: docker run -d --name test-22 --link=test-1:test-1 --link=test-2:test-2 --link=test-3:test-3 --link=test-4:test-4 --link=test-5:test-5 --link=test-6:test-6 --link=test-7:test-7 --link=test-8:test-8 --link=test-9:test-9 --link=test-10:test-10 --link=test-11:test-11 --link=test-12:test-12 --link=test-13:test-13 --link=test-14:test-14 --link=test-15:test-15 --link=test-16:test-16 --link=test-17:test-17 --link=test-18:test-18 --link=test-19:test-19 --link=test-20:test-20 --link=test-21:test-21  busybox sleep 10000
1420533832 86cb96143b4c07be28310bbf69031f1d341ca7fef1f84f94a906168e4de332ef
1420534243 docker ps -a | wc -l       23
```

Finally, this is a graph showing the correlation between the total number of links between existing containers and the time each iteration takes to complete (mainly running `docker ps`):

![image](https://cloud.githubusercontent.com/assets/27665/5626481/03876e36-9541-11e4-8336-b5b344cb2818.png)

This behavior severely impacts fig, and potentially or any other orchestration tool that lists the containers often (fig does WAY too often, but that's a separate issue).

Any clues on how to debug this would be appreciated. Following the code it would seem that is somehow related to graph traversal, but I haven't been able to verify this end.

**Comment by LK4D4** [comment]
Interesting, this is indeed can be related to graphdb. We have some profiling endpoints exposed when DEBUG env is set, you can use them with go tool pprof.

**Comment by LK4D4** [comment]
@tbatchelli Problem in line:
```
rows, err := db.conn.Query("SELECT entity_id, name FROM edge where parent_id = ?;", e.id)
```
in `pkg/graphdb`. Seems like sqlite is pretty slow :/

**Comment by tbatchelli** [comment]
Thanks for looking into it!

I am not familiar with the graph, but the line you point at should have at most O(n) complexity. In itself this doesn't look to be the root cause.

Looking at the function where this call is made ( https://github.com/docker/docker/blob/master/pkg/graphdb/graphdb.go#L475 ) it would seem that it recursively traverses the graph making the child call for each node, and that would explain why we'd be seeing this O(n^m). IF this call is made once per node in the graph (it seems it would), then the complexity would be O(n^2) and this would be consistent with what I am seeing, assuming it's an acyclic directed graph.

From this I understand that in the graph, containers are 'connected' by links? 

If all the assumptions above are correct, then I think this could be sorted by caching the traversals already done, as we'll be exploring the same subgraphs over and over, once per incoming link.

Does this sound right?

**Comment by LK4D4** [comment]
@tbatchelli Yes, sounds right. I missed recursive nature of that function even @erikh warned me about it.
Cache should help I think.

**Comment by tbatchelli** [comment]
My first attempt at fixing it didn't go down well.

Docker now is really fast, taking a few seconds to run the test above, and the tests pass, I can start containers and such, but fig fails, and it seems it is losing track of some of the containers in a fig file. I haven't been able to isolate the issue yet.

Here are my changes:

```diff
diff --git i/pkg/graphdb/graphdb.go w/pkg/graphdb/graphdb.go
index 6234203..a6d4d88 100644
--- i/pkg/graphdb/graphdb.go
+++ w/pkg/graphdb/graphdb.go
@@ -6,6 +6,7 @@ import (
 	"path"
 	"strings"
 	"sync"
+	log "github.com/Sirupsen/logrus"
 )
 
 const (
@@ -441,7 +442,17 @@ type WalkMeta struct {
 	Edge     *Edge
 }
 
+func entityInSlice (entity *Entity, entities []WalkMeta) bool {
+    for _, e := range entities {
+        if e.Parent.id == entity.id {
+            return true
+        }
+    }
+    return false
+}
+
 func (db *Database) children(e *Entity, name string, depth int, entities []WalkMeta) ([]WalkMeta, error) {
+	log.Infof("Getting Children")
 	if e == nil {
 		return entities, nil
 	}
@@ -458,6 +469,8 @@ func (db *Database) children(e *Entity, name string, depth int, entities []WalkM
 			return nil, err
 		}
 		child := &Entity{entityID}
+		
+		
 		edge := &Edge{
 			ParentID: e.id,
 			Name:     entityName,
@@ -471,8 +484,17 @@ func (db *Database) children(e *Entity, name string, depth int, entities []WalkM
 			Edge:     edge,
 		}
 
-		entities = append(entities, meta)
 
+		log.Infof("Number of entities to check %d, entity in list? %t ",
+			len(entities),
+		        entityInSlice(child, entities))
+
+		entities = append(entities, meta)
+		
+		if (entityInSlice(child,entities)){
+			return entities, nil
+		}
+		
 		if depth != 0 {
 			nDepth := depth
 			if depth != -1 {
```

**Comment by tbatchelli** [comment]
A late night of investigation has produced these new findings, and I would need confirmation wether this is the intended behavior or not:

1. The issue is not seem to be DB access. I have modified the code to preload the entities table in memory and then use the map as a lookup. This ensures the DB is accessed only once, and seems to  make no significant difference in the time it takes to complete the request.
2. `db.children` lists __every possible__ path from the start node to all the other nodes. If you have many services that are interconnected via links (a rather common scenario in micro-services) the number of potential paths can be huge. In a cluster with 20 microservices this number seemed to be around 150K.

For my use case things get exacerbated by `fig` because:

1. Fig makes a lot of repeated calls to `/containers`. Each is a call to `db.children`
2. Fig adds 3 actual links for each link in the fig.yml: <name>, <name_n> and <project>_<name>_<n>, making the issue with the number of path much much worse
3. Fig relies somehow on all the paths being present (the attempted fix above, which visits all the nodes but doesn't list all possible paths, reveals this issue). 


For example, if the following fig cluster is instantiated, the entities listed is the list below:
```yaml
a:
  image: busybox
  command: sleep 10000
b:
  image: busybox
  command: sleep 10000
  links:
    - a
c:
  image: busybox
  command: sleep 10000
  links:
    - a
    - b
d:
  image: busybox
  command: sleep 1000
  links:
    - a
    - b
    - c
e:
  image: busybox
  command: sleep 1000
  links:
    - c
```

```
time="2015-01-14T18:04:30Z" level="info" msg="### db.Children: entities=%!(EXTRA int=134)" 
time="2015-01-14T18:04:30Z" level="info" msg="Entities for /" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083802d0 Entity:0xc20836e3c0 FullPath:/slowbug_a_1 Edge:0xc208183800}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083802d0 Entity:0xc20836e560 FullPath:/slowbug_b_1 Edge:0xc208183ad0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836e560 Entity:0xc20836e780 FullPath:/slowbug_b_1/a Edge:0xc208183d10}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836e560 Entity:0xc20836e910 FullPath:/slowbug_b_1/a_1 Edge:0xc208183f80}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836e560 Entity:0xc20836f770 FullPath:/slowbug_b_1/slowbug_a_1 Edge:0xc20818d7a0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083802d0 Entity:0xc20836f8c0 FullPath:/slowbug_c_1 Edge:0xc20818c0c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836f8c0 Entity:0xc20836fc90 FullPath:/slowbug_c_1/a Edge:0xc20818c450}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836f8c0 Entity:0xc208354380 FullPath:/slowbug_c_1/a_1 Edge:0xc20818c840}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836f8c0 Entity:0xc208354560 FullPath:/slowbug_c_1/slowbug_a_1 Edge:0xc20818c9c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836f8c0 Entity:0xc2083556b0 FullPath:/slowbug_c_1/b Edge:0xc20818cd50}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083556b0 Entity:0xc208354730 FullPath:/slowbug_c_1/b/a Edge:0xc20818d2f0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083556b0 Entity:0xc2083548b0 FullPath:/slowbug_c_1/b/a_1 Edge:0xc20818d620}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083556b0 Entity:0xc2083549f0 FullPath:/slowbug_c_1/b/slowbug_a_1 Edge:0xc20818da10}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836f8c0 Entity:0xc208354c10 FullPath:/slowbug_c_1/b_1 Edge:0xc208154000}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208354c10 Entity:0xc208354e50 FullPath:/slowbug_c_1/b_1/a Edge:0xc2081552c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208354c10 Entity:0xc208354f80 FullPath:/slowbug_c_1/b_1/a_1 Edge:0xc208155c80}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208354c10 Entity:0xc208355110 FullPath:/slowbug_c_1/b_1/slowbug_a_1 Edge:0xc208154600}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20836f8c0 Entity:0xc208355270 FullPath:/slowbug_c_1/slowbug_b_1 Edge:0xc2081549f0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355270 Entity:0xc208355420 FullPath:/slowbug_c_1/slowbug_b_1/a Edge:0xc208155200}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355270 Entity:0xc208355630 FullPath:/slowbug_c_1/slowbug_b_1/a_1 Edge:0xc2081555f0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355270 Entity:0xc2083557b0 FullPath:/slowbug_c_1/slowbug_b_1/slowbug_a_1 Edge:0xc2081559b0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083802d0 Entity:0xc208355910 FullPath:/slowbug_e_1 Edge:0xc2081263f0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355910 Entity:0xc208355b60 FullPath:/slowbug_e_1/c Edge:0xc208127110}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355b60 Entity:0xc208355d80 FullPath:/slowbug_e_1/c/a Edge:0xc208127bf0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355b60 Entity:0xc208355ee0 FullPath:/slowbug_e_1/c/a_1 Edge:0xc2081261b0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355b60 Entity:0xc20833cb70 FullPath:/slowbug_e_1/c/slowbug_a_1 Edge:0xc2081264b0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355b60 Entity:0xc20833df40 FullPath:/slowbug_e_1/c/b Edge:0xc208126960}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833df40 Entity:0xc20833c230 FullPath:/slowbug_e_1/c/b/a Edge:0xc208126fc0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833df40 Entity:0xc20833c360 FullPath:/slowbug_e_1/c/b/a_1 Edge:0xc208127200}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833df40 Entity:0xc20833c4c0 FullPath:/slowbug_e_1/c/b/slowbug_a_1 Edge:0xc208127440}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355b60 Entity:0xc20833c640 FullPath:/slowbug_e_1/c/b_1 Edge:0xc208127b30}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833c640 Entity:0xc20833c800 FullPath:/slowbug_e_1/c/b_1/a Edge:0xc208127fb0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833c640 Entity:0xc20833c9b0 FullPath:/slowbug_e_1/c/b_1/a_1 Edge:0xc2080b30b0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833c640 Entity:0xc20833cb60 FullPath:/slowbug_e_1/c/b_1/slowbug_a_1 Edge:0xc2080b20f0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355b60 Entity:0xc20833cd50 FullPath:/slowbug_e_1/c/slowbug_b_1 Edge:0xc2080b3170}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833cd50 Entity:0xc20833cfe0 FullPath:/slowbug_e_1/c/slowbug_b_1/a Edge:0xc2080b3ec0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833cd50 Entity:0xc20833d190 FullPath:/slowbug_e_1/c/slowbug_b_1/a_1 Edge:0xc208410030}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833cd50 Entity:0xc20833d310 FullPath:/slowbug_e_1/c/slowbug_b_1/slowbug_a_1 Edge:0xc2084100c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355910 Entity:0xc20833d4e0 FullPath:/slowbug_e_1/c_1 Edge:0xc2084101b0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833d4e0 Entity:0xc20833d810 FullPath:/slowbug_e_1/c_1/a Edge:0xc208410270}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833d4e0 Entity:0xc20833d9a0 FullPath:/slowbug_e_1/c_1/a_1 Edge:0xc208410300}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833d4e0 Entity:0xc20833dad0 FullPath:/slowbug_e_1/c_1/slowbug_a_1 Edge:0xc208410390}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833d4e0 Entity:0xc20833dc30 FullPath:/slowbug_e_1/c_1/b Edge:0xc208410420}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833dc30 Entity:0xc20833de40 FullPath:/slowbug_e_1/c_1/b/a Edge:0xc2084104e0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833dc30 Entity:0xc20833df90 FullPath:/slowbug_e_1/c_1/b/a_1 Edge:0xc208410570}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833dc30 Entity:0xc20832f190 FullPath:/slowbug_e_1/c_1/b/slowbug_a_1 Edge:0xc208410600}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833d4e0 Entity:0xc20832e0a0 FullPath:/slowbug_e_1/c_1/b_1 Edge:0xc208410690}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832e0a0 Entity:0xc20832e320 FullPath:/slowbug_e_1/c_1/b_1/a Edge:0xc208410750}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832e0a0 Entity:0xc20832e4b0 FullPath:/slowbug_e_1/c_1/b_1/a_1 Edge:0xc2084107e0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832e0a0 Entity:0xc20832e6d0 FullPath:/slowbug_e_1/c_1/b_1/slowbug_a_1 Edge:0xc208410870}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20833d4e0 Entity:0xc20832e870 FullPath:/slowbug_e_1/c_1/slowbug_b_1 Edge:0xc208410900}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832e870 Entity:0xc20832eb30 FullPath:/slowbug_e_1/c_1/slowbug_b_1/a Edge:0xc2084109c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832e870 Entity:0xc20832ed00 FullPath:/slowbug_e_1/c_1/slowbug_b_1/a_1 Edge:0xc208410a50}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832e870 Entity:0xc20832ee80 FullPath:/slowbug_e_1/c_1/slowbug_b_1/slowbug_a_1 Edge:0xc208410ae0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc208355910 Entity:0xc20832efa0 FullPath:/slowbug_e_1/slowbug_c_1 Edge:0xc208410bd0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832efa0 Entity:0xc20832f1f0 FullPath:/slowbug_e_1/slowbug_c_1/a Edge:0xc208410c90}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832efa0 Entity:0xc20832f320 FullPath:/slowbug_e_1/slowbug_c_1/a_1 Edge:0xc208410d20}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832efa0 Entity:0xc20832f4f0 FullPath:/slowbug_e_1/slowbug_c_1/slowbug_a_1 Edge:0xc208410db0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832efa0 Entity:0xc20832f6a0 FullPath:/slowbug_e_1/slowbug_c_1/b Edge:0xc208410ea0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832f6a0 Entity:0xc20832f8a0 FullPath:/slowbug_e_1/slowbug_c_1/b/a Edge:0xc208410f60}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832f6a0 Entity:0xc20832f9d0 FullPath:/slowbug_e_1/slowbug_c_1/b/a_1 Edge:0xc208410ff0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832f6a0 Entity:0xc20832fb60 FullPath:/slowbug_e_1/slowbug_c_1/b/slowbug_a_1 Edge:0xc208411080}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832efa0 Entity:0xc20832fcb0 FullPath:/slowbug_e_1/slowbug_c_1/b_1 Edge:0xc208411170}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832fcb0 Entity:0xc20832fef0 FullPath:/slowbug_e_1/slowbug_c_1/b_1/a Edge:0xc208411230}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832fcb0 Entity:0xc2082f31f0 FullPath:/slowbug_e_1/slowbug_c_1/b_1/a_1 Edge:0xc2084112c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832fcb0 Entity:0xc2082f2140 FullPath:/slowbug_e_1/slowbug_c_1/b_1/slowbug_a_1 Edge:0xc208411350}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc20832efa0 Entity:0xc2082f22c0 FullPath:/slowbug_e_1/slowbug_c_1/slowbug_b_1 Edge:0xc208411440}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f22c0 Entity:0xc2082f2580 FullPath:/slowbug_e_1/slowbug_c_1/slowbug_b_1/a Edge:0xc208411560}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f22c0 Entity:0xc2082f2760 FullPath:/slowbug_e_1/slowbug_c_1/slowbug_b_1/a_1 Edge:0xc208411650}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f22c0 Entity:0xc2082f28c0 FullPath:/slowbug_e_1/slowbug_c_1/slowbug_b_1/slowbug_a_1 Edge:0xc208411740}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2083802d0 Entity:0xc2082f2a80 FullPath:/slowbug_d_1 Edge:0xc208411830}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082f2e40 FullPath:/slowbug_d_1/a Edge:0xc2084118f0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082f3110 FullPath:/slowbug_d_1/a_1 Edge:0xc208411980}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082f3320 FullPath:/slowbug_d_1/slowbug_a_1 Edge:0xc208411a10}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082f34c0 FullPath:/slowbug_d_1/b Edge:0xc208411aa0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f34c0 Entity:0xc2082f3790 FullPath:/slowbug_d_1/b/a Edge:0xc208411b60}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f34c0 Entity:0xc2082f39d0 FullPath:/slowbug_d_1/b/a_1 Edge:0xc208411bf0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f34c0 Entity:0xc2082f3b00 FullPath:/slowbug_d_1/b/slowbug_a_1 Edge:0xc208411c80}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082f3d00 FullPath:/slowbug_d_1/b_1 Edge:0xc208411d10}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f3d00 Entity:0xc2082f3ff0 FullPath:/slowbug_d_1/b_1/a Edge:0xc208411dd0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f3d00 Entity:0xc2082e01b0 FullPath:/slowbug_d_1/b_1/a_1 Edge:0xc208411e60}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f3d00 Entity:0xc2082e03b0 FullPath:/slowbug_d_1/b_1/slowbug_a_1 Edge:0xc208411ef0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082e06c0 FullPath:/slowbug_d_1/slowbug_b_1 Edge:0xc208411f80}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e06c0 Entity:0xc2082e0bb0 FullPath:/slowbug_d_1/slowbug_b_1/a Edge:0xc208432060}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e06c0 Entity:0xc2082e0e10 FullPath:/slowbug_d_1/slowbug_b_1/a_1 Edge:0xc2084320f0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e06c0 Entity:0xc2082e1050 FullPath:/slowbug_d_1/slowbug_b_1/slowbug_a_1 Edge:0xc208432180}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082e12e0 FullPath:/slowbug_d_1/c Edge:0xc208432270}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e12e0 Entity:0xc2082e1690 FullPath:/slowbug_d_1/c/a Edge:0xc208432330}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e12e0 Entity:0xc2082e1860 FullPath:/slowbug_d_1/c/a_1 Edge:0xc2084323c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e12e0 Entity:0xc2082e1b80 FullPath:/slowbug_d_1/c/slowbug_a_1 Edge:0xc208432450}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e12e0 Entity:0xc2082e1db0 FullPath:/slowbug_d_1/c/b Edge:0xc2084324e0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e1db0 Entity:0xc2082c8e20 FullPath:/slowbug_d_1/c/b/a Edge:0xc2084325a0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e1db0 Entity:0xc2082c8330 FullPath:/slowbug_d_1/c/b/a_1 Edge:0xc208432630}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e1db0 Entity:0xc2082c8580 FullPath:/slowbug_d_1/c/b/slowbug_a_1 Edge:0xc2084326c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e12e0 Entity:0xc2082c8830 FullPath:/slowbug_d_1/c/b_1 Edge:0xc208432750}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c8830 Entity:0xc2082c8b90 FullPath:/slowbug_d_1/c/b_1/a Edge:0xc208432810}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c8830 Entity:0xc2082c8db0 FullPath:/slowbug_d_1/c/b_1/a_1 Edge:0xc2084328a0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c8830 Entity:0xc2082c8f00 FullPath:/slowbug_d_1/c/b_1/slowbug_a_1 Edge:0xc208432930}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082e12e0 Entity:0xc2082c91f0 FullPath:/slowbug_d_1/c/slowbug_b_1 Edge:0xc2084329c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c91f0 Entity:0xc2082c9490 FullPath:/slowbug_d_1/c/slowbug_b_1/a Edge:0xc208432a80}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c91f0 Entity:0xc2082c95e0 FullPath:/slowbug_d_1/c/slowbug_b_1/a_1 Edge:0xc208432b10}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c91f0 Entity:0xc2082c9740 FullPath:/slowbug_d_1/c/slowbug_b_1/slowbug_a_1 Edge:0xc208432ba0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082c98d0 FullPath:/slowbug_d_1/c_1 Edge:0xc208432c90}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c98d0 Entity:0xc2082c9b90 FullPath:/slowbug_d_1/c_1/a Edge:0xc208432d50}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c98d0 Entity:0xc2082c9d10 FullPath:/slowbug_d_1/c_1/a_1 Edge:0xc208432de0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c98d0 Entity:0xc2082c9f20 FullPath:/slowbug_d_1/c_1/slowbug_a_1 Edge:0xc208432e70}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c98d0 Entity:0xc2082b3370 FullPath:/slowbug_d_1/c_1/b Edge:0xc208432f00}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3370 Entity:0xc2082b21f0 FullPath:/slowbug_d_1/c_1/b/a Edge:0xc208432fc0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3370 Entity:0xc2082b23b0 FullPath:/slowbug_d_1/c_1/b/a_1 Edge:0xc208433050}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3370 Entity:0xc2082b2560 FullPath:/slowbug_d_1/c_1/b/slowbug_a_1 Edge:0xc2084330e0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c98d0 Entity:0xc2082b27c0 FullPath:/slowbug_d_1/c_1/b_1 Edge:0xc208433170}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b27c0 Entity:0xc2082b2a20 FullPath:/slowbug_d_1/c_1/b_1/a Edge:0xc208433230}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b27c0 Entity:0xc2082b2c70 FullPath:/slowbug_d_1/c_1/b_1/a_1 Edge:0xc2084332c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b27c0 Entity:0xc2082b2e60 FullPath:/slowbug_d_1/c_1/b_1/slowbug_a_1 Edge:0xc208433350}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082c98d0 Entity:0xc2082b2fc0 FullPath:/slowbug_d_1/c_1/slowbug_b_1 Edge:0xc2084333e0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b2fc0 Entity:0xc2082b3270 FullPath:/slowbug_d_1/c_1/slowbug_b_1/a Edge:0xc2084334a0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b2fc0 Entity:0xc2082b3450 FullPath:/slowbug_d_1/c_1/slowbug_b_1/a_1 Edge:0xc208433530}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b2fc0 Entity:0xc2082b35a0 FullPath:/slowbug_d_1/c_1/slowbug_b_1/slowbug_a_1 Edge:0xc2084335c0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082f2a80 Entity:0xc2082b3710 FullPath:/slowbug_d_1/slowbug_c_1 Edge:0xc2084336b0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3710 Entity:0xc2082b3a00 FullPath:/slowbug_d_1/slowbug_c_1/a Edge:0xc208433770}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3710 Entity:0xc2082b3c20 FullPath:/slowbug_d_1/slowbug_c_1/a_1 Edge:0xc208433800}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3710 Entity:0xc2082b3dd0 FullPath:/slowbug_d_1/slowbug_c_1/slowbug_a_1 Edge:0xc208433890}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3710 Entity:0xc2082b3f30 FullPath:/slowbug_d_1/slowbug_c_1/b Edge:0xc208433980}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3f30 Entity:0xc2082a5a80 FullPath:/slowbug_d_1/slowbug_c_1/b/a Edge:0xc208433a40}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3f30 Entity:0xc2082a41a0 FullPath:/slowbug_d_1/slowbug_c_1/b/a_1 Edge:0xc208433ad0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3f30 Entity:0xc2082a4350 FullPath:/slowbug_d_1/slowbug_c_1/b/slowbug_a_1 Edge:0xc208433b60}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3710 Entity:0xc2082a4530 FullPath:/slowbug_d_1/slowbug_c_1/b_1 Edge:0xc208433c50}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082a4530 Entity:0xc2082a4840 FullPath:/slowbug_d_1/slowbug_c_1/b_1/a Edge:0xc208433d10}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082a4530 Entity:0xc2082a4970 FullPath:/slowbug_d_1/slowbug_c_1/b_1/a_1 Edge:0xc208433da0}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082a4530 Entity:0xc2082a4b70 FullPath:/slowbug_d_1/slowbug_c_1/b_1/slowbug_a_1 Edge:0xc208433e30}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082b3710 Entity:0xc2082a4d80 FullPath:/slowbug_d_1/slowbug_c_1/slowbug_b_1 Edge:0xc208433f20}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082a4d80 Entity:0xc2082a4fb0 FullPath:/slowbug_d_1/slowbug_c_1/slowbug_b_1/a Edge:0xc208558060}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082a4d80 Entity:0xc2082a51d0 FullPath:/slowbug_d_1/slowbug_c_1/slowbug_b_1/a_1 Edge:0xc208558150}" 
time="2015-01-14T18:04:30Z" level="info" msg="Ent: {Parent:0xc2082a4d80 Entity:0xc2082a53b0 FullPath:/slowbug_d_1/slowbug_c_1/slowbug_b_1/slowbug_a_1 Edge:0xc208558240}" 
```

**Comment by LK4D4** [comment]
Yup, you can also see enormous amount of `Names` in `/containers` output.

**Comment by tbatchelli** [comment]
So this is the expected behavior? If so, this can't be fixed without breaking the api...

**Comment by LK4D4** [comment]
@tbatchelli Yes, but I'm not sure that current api is useful at all.
